Add MapboxNavigationApp lifecycle

kmadsen opened the pull request:
### Description
<!--
Include issue references (e.g., fixes [#issue](link))
Include necessary implementation details (e.g. I opted to use this algorithm because ... and test it in this way ...).
-->

Resolves https://github.com/mapbox/mapbox-navigation-android/issues/5109

- [x] ~Create `:libnavigation-lifecycle:` module~ We decided to put this in libnativation-core
- [x] Port over `MapboxNavigationApp` `CarAppLifecycleOwner` `MapboxNavigationOwner` `MapboxNavigationObserver`
- [x] Port over tests

### Changelog
<!--
Include changelog entry (e.g. Fixed an unexpected change in recenter button when resuming the app.).
See https://github.com/mapbox/navigation-sdks/blob/main/documentation/android-changelog-guidelines.md.
You can remove the changelog block and add a `skip changelog` label, when applicable.
 -->
```
<changelog>Introduce `MapboxNavigationApp` and `MapboxNavigationObserver` to handle `MapboxNavigation` lifecycles. [#5112](https://github.com/mapbox/mapbox-navigation-android/pull/5112)</changelog>
```

kmadsen commented on libnavui-dropin/build.gradle:
```diff
@@ -45,6 +45,7 @@ dependencies {
     api project(":libnavui-tripprogress")
     api project(":libnavui-util")
     api project(":libnavui-voice")
+    api project(":libnavui-lifecycle")
```
We need to expose the ability to call `MapboxNavigationApp.setup` from the `Application` so we can register onto `registerActivityLifecycleCallbacks`

Zayankovsky commented on libnavui-lifecycle/src/main/java/com/mapbox/navigation/ui/lifecycle/internal/MapboxNavigationOwner.kt:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    fun unregister(mapboxNavigationObserver: MapboxNavigationObserver) {
+        mapboxNavigationObserver.onDetached(mapboxNavigation)
```
I don’t think we need to call onDetached, if mapboxNavigation is null. I would expect every onDetached call to follow exactly one onAttached call. And the argument should be non-null.

kmadsen replied:
Yeah I've gone back and forth on this but have also decided to remove the nullable. Over communicating the reasoning here: Having a nullable, supports construction initialization and removal. Removing the nullable, makes the construction initialization impossible. So the first approach opted for the extra signal. But am in the process of removing, because of the examples below

Supporting this pattern
``` kotlin
object : MapboxNavigationObserver {
  // FooReader is constructed with an enabled state
  val fooReader = FooReader()

  fun onAttached(mapboxNavigation: MapboxNavigation) {
    // -- snip -- //
  }

  fun onDetached(mapboxNavigation: MapboxNavigation?) {
    fooReader.disable()
  }
}
```

Removing, requires this pattern. Which I'm agreeing is better
``` kotlin
object : MapboxNavigationObserver {
  // FooReader is constructed with an enabled state
  late init var fooReader: FooReader

  fun onAttached(mapboxNavigation: MapboxNavigation) {
     fooReader = FooReader()
  }

  fun onDetached(mapboxNavigation: MapboxNavigation) {
    // This pattern will crash when MapboxNavigation is nullable
    fooReader.disable()
  }
}
```

Zayankovsky commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt:
```diff
@@ -61,46 +56,18 @@ class NavigationView : ConstraintLayout, LifecycleObserver {
         this.navigationViewOptions = navigationViewOptions
-        this.mapboxNavigation = navigationViewModel.mapboxNavigation
         initializeViewModels()
         this.lifeCycleOwner = context as? LifecycleOwner ?: throw LifeCycleOwnerNotFoundException()
```
Should this be replaced with lifecycleOwner from MapboxNavigationApp?

kmadsen replied:
NavigationView will have its own `lifecycleOwner` which represents the lifecycle of the `NavigationView`

So no this `lifecycleOwner` is unreleased to `MapboxNavigationApp`

kmadsen commented on CHANGELOG.md:
```diff
@@ -2,6 +2,23 @@
+#### Bug fixes and improvements
+- Add `:libnavui-lifecycle:` to share a `MapboxNavigation` instance between `:libnavui-dropin:` and `:libnavui-car:`. [#5112](https://github.com/mapbox/mapbox-navigation-android/pull/5112)
+
+### Mapbox dependencies
```
Incrementally adding the changelog seems like it'll always have a initial conflict. could add the next-version-beginning changelog when the previous is released

LukasPaczos replied:
We need to enable the union merge strategy for the changelog file fo this workflow to be reasonable. Refs https://github.com/mapbox/navigation-sdks/issues/1246.

kmadsen replied:
Yeah.. we haven't implemented the union merge yet so maybe I could just leave this PR out of it.

Or do we have already decided the strategy?

kmadsen replied:
i'm removing the changelog file changes and using the current system on the main line. 

when we deploy our new readme strategy this PR will abide by it

LukasPaczos commented on libnavui-lifecycle/src/main/java/com/mapbox/navigation/ui/lifecycle/MapboxNavigationApp.kt:
````diff
@@ -0,0 +1,135 @@
+ * activity that uses [MapboxNavigation].
+ *
+ * ```
+ * class MyActivity : ComponentActivity() {
+ *   override fun onCreate() {
+ *     MapboxNavigationApp
+ *       .setup(this)
+ *       .attach(this)
+ *   }
+ *   override fun onResume() {
+ *     MapboxNavigationApp
+ *       .registerObserver(locationObserver)
+ *   }
+ *   override fun onPause() {
+ *     MapboxNavigationApp
+ *       .unregisterObserver(locationObserver)
+ *   }
+ * }
````
This approach currently does not work because just calling `attach` does not register the `mapboxNavigationOwner.carAppLifecycleObserver` into `carAppLifecycleOwner` and `MapboxNavigation` is never delivered. This only happens in `attachAllActivities`.

kmadsen replied:
> This approach currently does not work because just calling attach does not register the mapboxNavigationOwner.carAppLifecycleObserver into carAppLifecycleOwner and MapboxNavigation is never delivered. 

What do you mean? `MapboxNavigationApp.attach` -> `CarAppLifecycleOwner.attach` ->  `startedReferenceCounter`

And the `startedReferenceCounter` is responsible for counting the "started" lifecycle events.

```
    fun attach(lifecycleOwner: LifecycleOwner) = apply {
        carAppLifecycleOwner.attach(lifecycleOwner)
    }

CarAppLifecycleOwner {
    fun attach(lifecycleOwner: LifecycleOwner) {
        logI(TAG, Message("attach"))
        lifecycleOwner.lifecycle.addObserver(startedReferenceCounter)
    }
```

LukasPaczos replied:
Problem is that `mapboxNavigationOwner.carAppLifecycleObserver` is never registered with `carAppLifecycleOwner` when calling just `attach`. So even though `carAppLifecycleOwner` is started, the `mapboxNavigationOwner` is never notified and the instance of `MapboxNavigation` is not created.

kmadsen replied:
This should be fixed. Any behavior issues we see can be shown in the `MapboxNavigationAppDelegateTest`

LukasPaczos commented on libnavui-lifecycle/src/main/java/com/mapbox/navigation/ui/lifecycle/internal/MapboxNavigationOwner.kt:
```diff
@@ -0,0 +1,73 @@
+            logI(TAG, Message("navigation.onStop"))
+            attached = false
+            services.forEach { it.onDetached(mapboxNavigation!!) }
+            MapboxNavigationProvider.destroy()
```
We're using the provider here but also storing the instance ourselves in `MapboxNavigationOwner#mapboxNavigation`. We either need to nullify that reference when we call `MapboxNavigationProvider.destroy()` or do not store the instance and use `MapboxNavigationProvider#retrieve` and work on local refs only.

kmadsen replied:
Am using `MapboxNavigationProvider.destroy()` to make it be backwards compatible. Otherwise, we could move the implementation of `MapboxNavigationApp` into `MapboxNavigationProvider`

Do you have a suggestion here?

kmadsen replied:
Didn't understand you earlier here. Suggestion was to set `mapboxNavigation = null`

That's fixed with a test covering it. Also there is a function to `MapboxNavigationApp`
``` kotlin
fun current(): MapboxNavigation? = mapboxNavigationOwner.current()
```

Found that it makes some code much simpler. Also prevents the need to store instances downstream everywhere. End of the day it is the same as `MapboxNavigationProvider#retrieve`, but with it we could deprecate `MapboxNavigationProvider` entirely.

LukasPaczos commented on libnavui-lifecycle/src/main/java/com/mapbox/navigation/ui/lifecycle/MapboxNavigationApp.kt:
```diff
@@ -0,0 +1,135 @@
+
+    /**
+     * Individually detach lifecycles from [MapboxNavigationApp].
+     * When all LifecycleOwners have been detached, this will cause all
+     * [MapboxNavigationObserver.onDetached].
```
This is a little bit inconvenient. If we wanted to use `MapboxNavigation` in an activity and add/remove listener in the `onAttached`/`onDetached` callbacks having another activity/fragment/viewmodel/AndroidAuto still working with `MapboxNavigation` would prevent us from getting `onDetached` callback and causing leaks. This means that we couldn't trust `onAttached`/`onDetached` lifecycle methods and still use local `onStart`/`onStop` instead, which defeats the purpose.

kmadsen replied:
Do you have a suggestion here?

The current idea is to separate the concerns. For example, your implementation of `MapboxNavigationObserver` does not know of the onStart, onStop events. So if you create a `RouteLineMbxObserver` it does not know of the lifecycle events.

But the Activity, is concerned with the lifecycle events. So your activity code looks like this (simple form) 
``` kotlin
MyActivity {
   onResume() { MapboxNavigationApp.register(routeLineMbxObserver) }
   onPause() { MapboxNavigationApp.unregister(routeLineMbxObserver) }
}

RouteLineMbxObserver {
  onAttach(mapboxNavigation) { renderRouteLine(mapboxNavigation) }
  onDetach(mapboxNavigation) { clearRouteLine(mapboxNavigation) }
}
```

And then can render the route line with arbitrary events

``` kotlin
MyRoutePreview {
   routeAvailable() { MapboxNavigationApp.register(routeLineMbxObserver) }
   userDoesntWantToSeeRoute() { MapboxNavigationApp.register(routeLineMbxObserver) }
}
```

LukasPaczos replied:
If there's another observer registered with `MapboxNavigationApp`, calling
```
onDestroy() { MapboxNavigationApp.detach(this) }
```
will not call
```
RouteLineMbxObserver#onDetach
```
defeating the purpose in my opinion, you will not have a clear understanding of when to unregistered your listeners from `MapboxNavigation` and you'll leak.

I think that the logic which calls `MapboxNavigationObserver#onDetach` shouldn't be tied to the main reference counter, but rather be invoked for each lifecycle owner separately when it's being detached or destroyed.

Only the `MapboxNavigation` instance should be tied to the main reference counter and destroyed when all lifecycles are finished.

kmadsen replied:
> will not call RouteLineMbxObserver#onDetach
> defeating the purpose in my opinion

I agree that, detaching should call onDetach. I'm pretty sure it does this, and we should be able to prove it with a unit test. Looking into this

kmadsen replied:
Excellent catch @LukasPaczos! Added a test and fixed it

LukasPaczos commented on libnavui-lifecycle/src/main/java/com/mapbox/navigation/ui/lifecycle/MapboxNavigationApp.kt:
```diff
@@ -0,0 +1,135 @@
+     *
+     * Must be called after [setup].
+     */
+    fun attachAllActivities() = apply {
+        val application = mapboxNavigationOwner.navigationOptions.applicationContext as Application
+        carAppLifecycleOwner.attachAllActivities(application)
+        carAppLifecycleOwner.lifecycle.addObserver(mapboxNavigationOwner.carAppLifecycleObserver)
+    }
```
Using this method forces us to use `disable` manually as well, otherwise, we never get `detached` callbacks, which means that we're not taking advantage of the actual lifecycle of the components that are working with the `MapboxNavigation`, they couldn't use `onAttached`/`onDetached` to manage registration/registration with observers.

I feel like this should be the "advanced user" method and general public should use `attach`/`detach` and pass lifecycle of components that actually interact with `MapboxNavigation`.

Any thoughts on this?

kmadsen replied:
yeah have been thinking about this. I was optimizing for "mapbox navigation ON" rather than "mapbox navigation OFF"

For example, make it easy to enable everywhere - and then custom effort to disable flows. That way it has "drop in" behavior. Default to ON. But then support granular lifecycles.

You could, disable everywhere and enable a single fragment with the `MapboxNavigationApp.attach`. But then, when you add a new fragment, you need to add `MapboxNavigationApp.attach` to make it work

You could, enable everywhere with `MapboxNavigationApp.attachAllActivities`, and then disable a specific flow or event with `MapboxNavigationApp.disable()`. When you add a new fragment, it will already have `MapboxNavigation` working

But you may be right, if you are a application like AirBnB. 95% of your application will be not driving, and then "navigating to your host" will want to enable and disable a lifecycle. So I do agree with you, but what does it change?

LukasPaczos replied:
Right, in your big app which has a sub-feature of navigation you want the lifecycle of `MapboxNavigation` be scoped to the lifecycles that actually use it and take advantage of `MapboxNavigationObserver` to register and unregistered your listeners so that you don't leak when moving between screens and features.

More of why this currently wouldn't work is in https://github.com/mapbox/mapbox-navigation-android/pull/5112#discussion_r756784503.

kmadsen replied:
This has been updated and should be working.

LukasPaczos commented on libnavui-lifecycle/src/main/java/com/mapbox/navigation/ui/lifecycle/MapboxNavigationObserver.kt:
```diff
@@ -0,0 +1,65 @@
+ *     mapboxNavigation.registerLocationObserver(locationObserver)
+ *   }
+ *
+ *   override fun onDetached(mapboxNavigation: MapboxNavigation?) {
+ *     mapboxNavigation?.unregisterLocationObserver(locationObserver)
+ *   }
```
```suggestion
 *   override fun onDetached(mapboxNavigation: MapboxNavigation) {
 *     mapboxNavigation.unregisterLocationObserver(locationObserver)
 *   }
```

kmadsen replied:
Yes!

LukasPaczos commented on libnavui-lifecycle/src/main/java/com/mapbox/navigation/ui/lifecycle/MapboxNavigationObserver.kt:
```diff
@@ -0,0 +1,65 @@
+ *   }
+ * }
+ *
+ * class LocationNavigationObserver : MapboxNavigationObserver {
```
```suggestion
 * class MyMapboxNavigationObserver : MapboxNavigationObserver {
```

LukasPaczos commented on libnavui-lifecycle/src/main/java/com/mapbox/navigation/ui/lifecycle/MapboxNavigationObserver.kt:
````diff
@@ -0,0 +1,65 @@
+ * Example of observing locations with a view model
+ * ```
+ * class MyViewModel : ViewModel() {
+ *   private val locationObserver = LocationObserver()
+ *
+ *   val location: LiveData<Location> = locationObserver.location.asLiveData()
````
```suggestion
 *   private val locationObserver = LocationObserver()
 *
 *   val location: LiveData<Location> = locationObserver.location.asLiveData()
 *
 * val myMapboxNavigationObserver = MyMapboxNavigationObserver()
```

LukasPaczos replied:
This might also need a cleanup of how location actually gets to `val location: LiveData<Location>`

kmadsen replied:
Yeah we may also build some tools on top of the callbacks to make it simpler. Goal is to make it so we can [simplify customer code like this](https://github.com/mapbox/mapbox-navigation-android-examples/blob/2878edfb3fcbac0a8db8266f0d5f156267d2ccf5/app/src/main/java/com/mapbox/navigation/examples/basics/ShowCurrentLocationActivity.kt#L51-L86) 

Here's the version we're using in android auto [CarAppLocationImpl](https://github.com/mapbox/mapbox-navigation-android-examples/blob/main/android-auto/src/main/java/com/mapbox/androidauto/navigation/location/impl/CarAppLocationImpl.kt)

LukasPaczos commented on libnavui-lifecycle/src/main/java/com/mapbox/navigation/ui/lifecycle/MapboxNavigationObserver.kt:
```diff
@@ -0,0 +1,65 @@
+    fun onAttached(mapboxNavigation: MapboxNavigation)
+
+    /**
+     * Is called when all activities are destroyed and the car session is destroyed.
```
```suggestion
     * Is called when all activities or services are destroyed/detached and there are no consumers of `MapboxNavigation`.
```

LukasPaczos replied:
But as mentioned above, this should be granular per each lifecycle in my opinion.

kmadsen replied:
Updated this to be more about `LifecycleOwner`. Services don't have a lifecycle, but you can still create your own LifecycleOwner (which is what androidx.car lib did). That way it is also granular.

I don't see harm in having the `attachAllActivities` available to developers. But maybe we should organize the comments or code to suggest using "attach/detach" functions.

LukasPaczos commented on settings.gradle:
```diff
@@ -19,5 +19,6 @@ include ':examples',
         ':libnavui-resources',
         ':libnavui-voice',
         ':libnavui-speedlimit',
+        ':libnavui-lifecycle',
```
```suggestion
        ':libnavigation-lifecycle',
```
This is more of a core module, isn't it?

kmadsen replied:
yeah i was going back and forth here 🤔 .. from comments above, am even considering moving to the same place as `MapboxNavigationProvider`

kmadsen replied:
Moved it

LukasPaczos commented on Makefile:
```diff
@@ -17,6 +17,7 @@ libnavui-resources \
 libnavui-voice \
 libnavigation-android \
 libnavui-speedlimit \
+libnavui-lifecycle \
```
Move to core modules.

Zayankovsky commented on libnavui-lifecycle/build.gradle:
```diff
@@ -0,0 +1,39 @@
+apply plugin: 'com.android.library'
+apply plugin: 'kotlin-android'
+apply plugin: 'kotlin-android-extensions'
```
We don't use anything from this plugin and it is deprecated anyway.

kmadsen replied:
copy paste from kotlin-android-extensions `libnavui-speedlimit` so we should delete there too

kmadsen replied:
Removing other places https://github.com/mapbox/mapbox-navigation-android/pull/5183

LukasPaczos commented on libnavigation-lifecycle/src/main/java/com/mapbox/navigation/lifecycle/internal/CarAppLifecycleOwner.kt:
```diff
@@ -0,0 +1,177 @@
+    }
+
+    private companion object {
+        private val TAG = Tag("CarAppLifecycleOwner")
```
```suggestion
        private val TAG = Tag("MbxCarAppLifecycleOwner")
```

LukasPaczos reviewed (commented):
Looking better and better, thanks for addressing the comments!

There's still one problem that I'm noticing - the consumer gets an `onDetached` callback only when the lifecycle is actually destroyed. This works well for preventing pure memory leaks but does not for a typical use case which is about avoiding updates when the lifecycle is stopped.

If you are using some of the observers from `MapboxNavigation` to update your UI, you don't want to do that when lifecycle of your activity or fragment is `stopped` and the current `MapboxNavigationObserver` does not help with that. You still need to override `Activity#onStop`/`Activity#onStart` and unsubscribe/resubscribe the observers next to doing the same in `MapboxNavigationObserver` callbacks.

Do you think the `MapboxNavigationObserver` could react to the lifecycle being stopped and started while only `MapboxNavigation` destruction being tied to all lifecycles being detached or destroyed?

kmadsen commented on libnavigation-lifecycle/build.gradle:
```diff
@@ -0,0 +1,39 @@
+}
+
+dependencies {
+    implementation project(":libnavigation-core")
```
> Alternatively we can leave it as is for now while it's experimental, and move it to the core module when we think it's ready to become stable.

I'm in favor of this approach only because the jetpack lifecycle would need to be added as an `api` dependency in `core`. Maybe that's fine? But as a separate module we won't have to worry about it yet.

LukasPaczos replied:
I think that's fine, if we want to take`LifecycleOwner`s as arguments, we need to provide this dependency.

The only question is whether we expect this approach to be the recommended way of accessing the `MapboxNavigation` object instead of `MapboxNavigationProvider` and the answer seems to be yes, so I think it should live in `core` so that developers don't need to specify another dependency to get it.

kmadsen replied:
Ok sounds good to me, moved to core 👍

kmadsen commented on libnavigation-lifecycle/src/main/java/com/mapbox/navigation/lifecycle/MapboxNavigationApp.kt:
```diff
@@ -0,0 +1,140 @@
+    /**
+     * Call [MapboxNavigationApp.setup] to provide the application with [NavigationOptions].
+     */
+    fun setup(navigationOptions: NavigationOptions) = apply {
```
difficult choice 🤔. `enable` is not perfect either because `attach*` is needed in order to enable the lifecycles.

LukasPaczos replied:
Naming... I'm fine with whatever you think is best!

kmadsen commented:
> Looking better and better, thanks for addressing the comments!
>
> There's still one problem that I'm noticing - the consumer gets an onDetached callback only when the lifecycle is actually destroyed. This works well for preventing pure memory leaks but does not for a typical use case which is about avoiding updates when the lifecycle is stopped.
>
> If you are using some of the observers from MapboxNavigation to update your UI, you don't want to do that when lifecycle of your activity or fragment is stopped and the current MapboxNavigationObserver does not help with that. You still need to override Activity#onStop/Activity#onStart and unsubscribe/resubscribe the observers next to doing the same in MapboxNavigationObserver callbacks.
>
> Do you think the MapboxNavigationObserver could react to the lifecycle being stopped and started while only MapboxNavigation destruction being tied to all lifecycles being detached or destroyed?

I'm not positive I'm catching what you mean, but adding some examples based on what I think you're saying.

You can control the your granular lifecycles like this https://github.com/mapbox/mapbox-navigation-android/pull/5112#discussion_r756535669
``` kotlin
MyFragment or MyActivity {
   // Adds this fragment lifecycle, MapboxNavigation will be created
   onCreate() { MapboxNavigationApp.attach(this) }
   
   // routeLineMbxObserver.onAttached will be called
   onResume() { MapboxNavigationApp.register(routeLineMbxObserver) }

   // routeLineMbxObserver.onDetached will be called
   onPause() { MapboxNavigationApp.unregister(routeLineMbxObserver) }

   // this is not required because lifecycle destroy events are terminal events.
   // But routeLineMbxObserver.onDetached would be called, even if you do not unregister onPause in this isolated case.
   // BUT Unregister onPause should still be required, because `MapboxNavigationApp` may have a different
   // Activity lifecycle attached and that would cause MyFragment to potentially leak.
   onDestroy() { MapboxNavigationApp.detach(this) }
}
```
So this means onDetached will be called when the lifecycle is destroyed OR when you unregister the observer.

I've been considering some wrapper extensions to handle all these extra calls, but they'll take more thought. But this is what I think would be a nice extension function. What I was investigating, is a "setOf<LifecycleOwner>", cause we may need a counter for each lifecycle owner attached. 
``` kotlin
/** 
 * automatically subscribes your observer when the lifecycle is in the resumed state,
 * unsubscribes the observer otherwise
 **/
fun MapboxNavigationApp.attachOnResumed(lifecycleOwner: LifecycleOwner, mapboxNavigationObserver: MapboxNavigationObserver) { ...
```

LukasPaczos commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/lifecycle/CarAppLifecycleOwner.kt:
```diff
@@ -0,0 +1,176 @@
+    internal val startedReferenceCounter = object : DefaultLifecycleObserver {
+        override fun onCreate(owner: LifecycleOwner) {
+            lifecycleCreated++
+            logger.i(TAG, Message("car onCreate"))
```
```suggestion
            logger.i(TAG, Message("lifecycle owner ($owner) onCreate"))
```
let's make this log clearer, same everywhere below. Having access to string representation or hash code will help us debug.

kmadsen replied:
yeah, it used to only allow a "car" lifecycle to attach. BUt now it's any LifecycleOwner. So this was stale, nice catch!

LukasPaczos commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/lifecycle/MapboxNavigationOwner.kt:
```diff
@@ -0,0 +1,75 @@
+    fun current(): MapboxNavigation? = mapboxNavigation
+
+    private companion object {
+        private val TAG = Tag("MapboxNavigationOwner")
```
```suggestion
        private val TAG = Tag("MbxNavigationOwner")
```

LukasPaczos reviewed (approved):
:ship: Dropping some smaller comments below, some we might want to address before merging and let's iterate on the rest.

Nice work!

LukasPaczos commented on libnavigation-lifecycle/src/main/java/com/mapbox/navigation/lifecycle/MapboxNavigationObserver.kt:
```diff
@@ -0,0 +1,75 @@
+     * This is called after all [LifecycleOwner]s that have been attached to
+     * [MapboxNavigationApp.attach] have moved to the [Lifecycle.State.CREATED] or
+     * [Lifecycle.State.DESTROYED] state. Calling [MapboxNavigationApp.detach] will cause all
+     * registered observers to detach when the attached lifecycle owner state condition is met.
```
This comment still stands.

kmadsen replied:
Yeah agree, re-wrote it again ❤️

LukasPaczos commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/lifecycle/MapboxNavigationObserver.kt:
```diff
@@ -0,0 +1,75 @@
+     * will only be called if [onAttached] has been called. The [mapboxNavigation] instance will
+     * be the same instance as the one passed to the [onAttached].
+     *
+     * This is called after all [LifecycleOwner]s that have been attached to
```
Do _all_ owners need to be started or rather _any_ owner?

kmadsen replied:
Yeah it's all lifecycle owners need to be stopped. Allowing for car and app. 

Future iteration here of `MapboxNavigationApp`, or an extension of the singleton, will be to hook a MapboxNavigationObserver to a specific LifecycleOwner.

kmadsen replied:
Updating the description. But may need to add something to address this expectation

kmadsen commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/lifecycle/MapboxNavigationOwner.kt:
```diff
@@ -0,0 +1,75 @@
+            check(!MapboxNavigationProvider.isCreated()) {
+                "MapboxNavigation should only be created by the MapboxNavigationOwner"
+            }
+            check(this@MapboxNavigationOwner::navigationOptions.isInitialized) {
+                "You must set the NavigationOptions with MapboxNavigationApp.setup"
+            }
```
It was necessary earlier on, we had a strict "setup first" requirement. But have been iterating since then, and this check will never actually happen. I'm replacing it with a test to demonstrate that it works.

```
    @Test
    fun `verify setup can be called after LifecycleOwner is started`() {
        val firstObserver = mockk<MapboxNavigationObserver>(relaxUnitFun = true)
        val secondObserver = mockk<MapboxNavigationObserver>(relaxUnitFun = true)
        mapboxNavigationApp.registerObserver(firstObserver)
        mapboxNavigationApp.registerObserver(secondObserver)

        val testLifecycleOwner = CarAppLifecycleOwnerTest.TestLifecycleOwner()
        mapboxNavigationApp.attach(testLifecycleOwner)

        testLifecycleOwner.lifecycleRegistry.currentState = Lifecycle.State.RESUMED
        mapboxNavigationApp.setup(navigationOptions)

        verify(exactly = 1) { firstObserver.onAttached(any()) }
        verify(exactly = 1) { secondObserver.onAttached(any()) }
        verify(exactly = 0) { firstObserver.onDetached(any()) }
        verify(exactly = 0) { secondObserver.onDetached(any()) }
    }
```

kmadsen commented on libnavigation-core/build.gradle:
```diff
@@ -54,6 +54,8 @@ dependencies {
     implementation dependenciesList.coroutinesAndroid
     implementation dependenciesList.androidStartup
 
+    implementation dependenciesList.androidXLifecycleRuntime
```
Yep we discussed this offline, thanks for catching!

Zayankovsky commented on LICENSE.md:
```diff
@@ -321,14 +321,26 @@ License: [The Apache Software License, Version 2.0](http://www.apache.org/licens
 
 ===========================================================================
 
+Mapbox Navigation uses portions of the Android Lifecycle Kotlin Extensions (Kotlin extensions for 'lifecycle' artifact).
+URL: [https://developer.android.com/jetpack/androidx/releases/lifecycle#2.4.0](https://developer.android.com/jetpack/androidx/releases/lifecycle#2.4.0)
+License: [The Apache Software License, Version 2.0](http://www.apache.org/licenses/LICENSE-2.0.txt)
```
Do we need this? We removed Kotlin Android Extensions recently. Also this seems to duplicate the next section.

kmadsen replied:
These were automatically generated from `make license` and `make license-verification`

Also, maybe you're mistaken here because it's "Android Lifecycle Kotlin Extensions". Not the "Kotlin Android Extensions"

Zayankovsky replied:
> Also, maybe you're mistaken here because it's "Android Lifecycle Kotlin Extensions". Not the "Kotlin Android Extensions"

Yep, I was wrong.

Zayankovsky reviewed (commented):
I'd like to note that when attaching lifecycle owners we don't track which of them are changing configurations unlike when attaching all activities. So `MapboxNavigation` will be recreated after configuration change, if I'm not mistaken. Is this something that we want to improve in the future? E.g. we could check whether the attached lifecycle owner is an instance of `Activity` or `Fragment` and have additional checks when it is destroyed.

kmadsen commented:
> I'd like to note that when attaching lifecycle owners we don't track which of them are changing configurations unlike when attaching all activities. So MapboxNavigation will be recreated after configuration change, if I'm not mistaken. Is this something that we want to improve in the future? E.g. we could check whether the attached lifecycle owner is an instance of Activity or Fragment and have additional checks when it is destroyed.

You're right, changing configurations is handled by `attachAllActivities` but not by the individual `attach` `detach` functions. That's an issue, looking at how we should handle it

kmadsen commented:
> > I'd like to note that when attaching lifecycle owners we don't track which of them are changing configurations unlike when attaching all activities. So MapboxNavigation will be recreated after configuration change, if I'm not mistaken. Is this something that we want to improve in the future? E.g. we could check whether the attached lifecycle owner is an instance of Activity or Fragment and have additional checks when it is destroyed.

> You're right, changing configurations is handled by attachAllActivities but not by the individual attach detach functions. That's an issue, looking at how we should handle it

I've added checks for a ComponentActivity (base activity that is a LifecycleOwner). This does not handle Fragment lifecycles. [Fragments have a few solutions](https://stackoverflow.com/questions/34855464/fragment-recreation-on-configuration-change), so I think we should leave the solution out of the framework for now. 

A couple tests that verify this is working
``` kotlin
fun `startedReferenceCounter verify orientation switching does not stop the app lifecycle`() {
fun `startedReferenceCounter verify backgrounded orientation switching does not stop the app lifecycle`() {
```

I've started to make improvements to `AppLifecycleActivity`, but that Activity is a derivative example from `AlternativeRouteActivity` which also does not handle configuration changes properly. Example issue is that, in order to prevent the `MapboxNavigationApp.setup` call from destroying the `MapboxNavigation` instance, we need to check if the lifecycle is undergoing a configuration change. 

I think these changes are worth further discussion, and there are multiple approaches we can consider. So I'll merge this one and immediately start the follow up discussions.
